File: 0070-climbing-stairs/0070-climbing-stairs.java
class Solution {
    public int climbStairs(int n) {
        int[] dp = new int[n+1];
        for(int i=0; i<n+1; i++) dp[i] = -1;
        return climbStairsHelper(n, 0, dp);
    }
    private int climbStairsHelper(int n, int curr, int[] dp){
        if(curr==n) return 1; // reached
        if(curr>n) return 0;
        if(dp[curr]!=-1) return dp[curr];
        return dp[curr] = climbStairsHelper(n, curr+1, dp)+climbStairsHelper(n, curr+2, dp); 
    }
}

File: 0070-climbing-stairs/0070-climbing-stairs.cpp
class Solution {
public:
    int dp[46];
    int climbStairs(int n) {
        if(n<0) return 0;
        if(n==0) return 1; 
        if(dp[n]!=0) return dp[n];   
        return dp[n] = climbStairs(n-1)+ climbStairs(n-2);
    }
};